src/assembler.jl

```diff
@@ -2,58 +2,82 @@ abstract type AbstractAssembler end
 abstract type AbstractCSCAssembler <: AbstractAssembler end
 
 """
-This assembler creates a COO (**coo**rdinate format) representation of a sparse matrix during
-assembly and converts it into a `SparseMatrixCSC{T}` on finalization.
+    COOAssembler
+
+This assembler creates a COO (**coo**rdinate format) representation of a sparse matrix
+during assembly and converts it into a `SparseMatrixCSC{T}` on finalization.
 """
-struct COOAssembler{T} # <: AbstractAssembler
-    I::Vector{Int}
-    J::Vector{Int}
-    V::Vector{T}
+struct COOAssembler{Tv, Ti} # <: AbstractAssembler
+    nrows::Int
+    ncols::Int
+    f::Vector{Tv}
+    I::Vector{Ti}
+    J::Vector{Ti}
+    V::Vector{Tv}
 end
 
-COOAssembler(N) = COOAssembler(Float64, N)
-
-function COOAssembler(::Type{T}, N) where T
+function COOAssembler{Tv, Ti}(nrows::Int, ncols::Int; sizehint::Int = 0) where {Tv, Ti}
     I = Int[]
     J = Int[]
-    V = T[]
-    sizehint!(I, N)
-    sizehint!(J, N)
-    sizehint!(V, N)
-
-    COOAssembler(I, J, V)
+    V = Tv[]
+    sizehint!(I, sizehint)
+    sizehint!(J, sizehint)
+    sizehint!(V, sizehint)
+    f = Tv[]
+    return COOAssembler{Tv, Ti}(nrows, ncols, f, I, J, V)
 end
 
 """
-    start_assemble([N=0]) -> COOAssembler
+    start_assemble(; sizehint::Int = 0) -> COOAssembler
+    start_assemble(nrows::Int, ncols::Int; sizehint::Int = 0) -> COOAssembler
 
-Create an `Assembler` object which can be used to assemble element contributions to the
-global sparse matrix. Use [`assemble!`](@ref) for each element, and [`finish_assemble`](@ref),
-to finalize the assembly and return the sparse matrix.
+Create an `COOAssembler` which can be used to assemble element contributions to a global
+sparse matrix and vector. `nrows` is the number of rows in the final matrix and `ncols` the
+number of columns. If `nrows` and `ncols` are not passed they are inferred from the
+maximum indices added to the assembler during assembly. `sizehint` is a hint for how many
+entries in total will be added to the assembler and can be passed to optimize for
+allocations.
 
-Note that giving a sparse matrix as input can be more efficient. See below and
-as described in the [manual](@ref man-assembly).
+Use [`assemble!`](@ref) to insert element contributions, and [`finish_assemble`](@ref), to
+finalize the assembly and return the sparse matrix (and optionally vector).
+
+Note that allocating a sparse matrix and assemble into it is generally preferred. See below
+and the [manual section on assembly](@ref man-assembly).
 
 !!! note
     When the same matrix pattern is used multiple times (for e.g. multiple time steps or
     Newton iterations) it is more efficient to create the sparse matrix **once** and reuse
     the same pattern. See the [manual section](@ref man-assembly) on assembly.
 """
-function start_assemble(N::Int=0)
-    return start_assemble(Float64, N)
+function start_assemble(nrows::Int, ncols::Int; sizehint::Int = 0)
```

Comment:
Should we include some dummy data type, e.g. `COOMatrixInfo`, to make the interface more symmetric (cc @KnutAM )? I left the interface intact as it was in the last PR, because the last PR was about refcatoring internals only.

Reply:
To encode the size or what?

Reply:
I don't think that makes it more symmetric, since we do pass the actual matrix [and vector] in the other cases, which you can refer to directly after assembly. That would not be the case here AFAIU. 

So I guess its ok as it is, but I think the best would be to have `start_assemble(data_to_be_filled...)::AbstractAssembler`, and that users who wants the `COOAssembler` should simply call `assembler = COOAssembler(args...; kwargs...)`